com.unity.ml-agents.extensions/Runtime/Input/InputActuatorComponent.cs

```diff
@@ -0,0 +1,376 @@
+        public override ActionSpec ActionSpec => m_ActionSpec;
+
+        /// <summary>
+        /// Creates a virtual <see cref="InputDevice"/> with a layout that is based on the <see cref="InputAction"/>s
+        /// in the <see cref="InputActionMap"/> specified by <see cref="PlayerInput.defaultActionMap"/>.
+        /// </summary>
+        /// <param name="layoutName">The custom layout name to use for the <see cref="InputDevice"/> to create.</param>
+        /// <param name="interfaceName">The made up interface name we are using for these virtual devices.  Right now
+        /// it is equal to <see cref="mlAgentsDeviceName"/> + <see cref="BehaviorParameters.BehaviorName"/></param>
+        /// <param name="localId"></param>
+        /// <returns></returns>
+        internal static InputDevice CreateDevice(string layoutName, string interfaceName, uint localId)
+        {
+            // See if our device layout was already registered.
+            var existingLayout = InputSystem.TryFindMatchingLayout(new InputDeviceDescription
+            {
+                interfaceName = interfaceName
+            });
+
+            // Load the device layout based on the controls we created previously.
+            if (string.IsNullOrEmpty(existingLayout))
+            {
+                InputSystem.RegisterLayoutMatcher(layoutName, new InputDeviceMatcher()
+                    .WithInterface($"^({interfaceName}[0-9]*)"));
+            }
+
+            // Actually create the device instance.
+            var device = InputSystem.AddDevice(
+                new InputDeviceDescription
+                {
+                    interfaceName = interfaceName,
+                    serial = $"{localId}"
+                }
+            );
+            return device;
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="defaultMap"></param>
+        /// <param name="layoutName"></param>
+        /// <param name="localId"></param>
+        /// <param name="behaviorParameters"></param>
+        /// <returns></returns>
+        internal static IActuator[] GenerateActionActuatorsFromAsset(
+            InputActionMap defaultMap,
+            string layoutName,
+            uint localId,
+            BehaviorParameters behaviorParameters)
+        {
+            // TODO does this need to change based on the action map we use?
+            var builder = new InputControlLayout.Builder()
+                .WithName(layoutName)
+                .WithFormat(mlAgentsLayoutFormat);
+
+            var actuators = new List<IActuator>();
+
+            foreach (var action in defaultMap)
+            {
+                var actionLayout = InputSystem.LoadLayout(action.expectedControlType);
+
+                builder.AddControl(action.name)
+                    .WithLayout(action.expectedControlType)
+                    .WithFormat(actionLayout.stateFormat)
+                    .WithUsages($"MLAgentPlayer-{localId}");
```

Comment:
What's the usages on controls used for?

Reply:
ah, this was an experiment when trying to map devices to instances of InputActionAssets, I can remove it.